File: leetcode/2373/largest_local_values_matrix.cpp
#include<vector>

using namespace std;

class Solution {

public:
    int largestGrid(vector<vector<int>>& grid) {
      int max_elem = grid[0][0];
      for (int i = 0; i < grid.size(); i++) {
        auto curr_max = max_element(grid[i].begin(), grid[i].end());
        max_elem = max(max_elem, *curr_max);
      }
      return max_elem;
    }

    vector<vector<int>> largestLocal(vector<vector<int>>& grid) {
        int n = grid.size() - 2;
        vector<vector<int>> result;

        for (int i = 0; i < n; i++) {
          vector<int> row;
          for (int j = 0; j < n; j++) {
            vector<vector<int>> localGrid;
            localGrid.push_back(vector<int>(grid[i].begin()+j, grid[i].begin()+j+3));
            localGrid.push_back(vector<int>(grid[i+1].begin()+j, grid[i+1].begin()+j+3));
            localGrid.push_back(vector<int>(grid[i+2].begin()+j, grid[i+2].begin()+j+3));
            row.push_back(largestGrid(localGrid));
          };
          result.push_back(row);
        };
        return result;
    };
};

File: leetcode/2373/largest_local_values_matrix.py
class Solution:
    def largestLocal(self, grid: List[List[int]]) -> List[List[int]]:
        i,j = 0,0
        n = len(grid) - 2
        result = []
        for i in range(n):
            row = []
            for j in range(n):
                subGrid = [
                    grid[i][j:j+3],
                    grid[i+1][j:j+3],
                    grid[i+2][j:j+3],
                ]
                row.append(self.largest3x3Grid(subGrid))
            result.append(row) 
        return result
    
    def largest3x3Grid(self, grid: List[List[int]]) -> int:
        maxLi = 0
        for li in grid:
            maxLi = max(maxLi, max(li))

        return maxLi
